module/VuFind/src/VuFind/Tags/TagsService.php

```diff
@@ -87,11 +95,76 @@ public function parse($tags)
      *
      * @return void
      */
-    public function addTagsToRecord(RecordDriver $driver, UserEntityInterface $user, array $tags): void
+    public function linkTagsToRecord(RecordDriver $driver, UserEntityInterface $user, array $tags): void
     {
         $resource = $this->resourcePopulator->getOrCreateResourceForDriver($driver);
         foreach ($tags as $tag) {
-            $resource->addTag($tag, $user);
+            $this->linkTagToResource($tag, $resource, $user);
+        }
+    }
+
+    /**
+     * Unlink a tag from a resource object.
+     *
+     * @param string                           $tagText      Text of tag to link (empty strings will be ignored)
+     * @param ResourceEntityInterface|int      $resourceOrId Resource entity or ID to link
+     * @param UserEntityInterface|int          $userOrId     Owner of tag link
+     * @param null|UserListEntityInterface|int $listOrId     Optional list (omit to tag at resource level)
+     *
+     * @return void
+     */
+    public function linkTagToResource(
+        string $tagText,
+        ResourceEntityInterface|int $resourceOrId,
+        UserEntityInterface|int $userOrId,
+        UserListEntityInterface|int|null $listOrId = null
+    ): void {
+        $trimmedTagText = trim($tagText);
+        if (strlen($trimmedTagText) > 0) {
```

Comment:
Nit: I'd compare with empty string. Even if it doesn't really make a difference here, we don't need the length.